File: boj/gold/5972.py
import sys
import heapq

input = sys.stdin.readline

n, m = map(int, input().split())
graph = [[] for _ in range(n + 1)]
for _ in range(m):
    a, b, c = map(int, input().split())
    graph[a].append((c, b))
    graph[b].append((c, a))

dist = [1e9] * (n + 1)
dist[1] = 0
pq = [] # ( cost, next )
heapq.heappush(pq, (dist[1], 1))

while len(pq) != 0:
    cost, x = heapq.heappop(pq)

    if cost > dist[x]:
        continue

    for (nxtCost, nxt) in graph[x]:
        if dist[nxt] > dist[x] + nxtCost:
            dist[nxt] = dist[x] + nxtCost
            heapq.heappush(pq, (dist[nxt], nxt))

print(dist[n])

File: boj/gold/5972.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    cin.tie(nullptr);
    cout.tie(nullptr);
    ios_base::sync_with_stdio(false);

    int n, m;
    cin >> n >> m;
    vector<pair<int, int>> graph[n + 1]; // { cost, destination }
    while (m--) {
        int a, b, c;
        cin >> a >> b >> c;
        graph[a].push_back({ c, b });
        graph[b].push_back({ c, a });
    }

    int dist[n + 1];
    fill(dist, dist + n + 1, 1e9);
    dist[1] = 0;

    priority_queue<
        pair<int, int>,
        vector<pair<int, int>>,
        greater<>
    > pq;
    pq.push({ dist[1], 1 });

    while (!pq.empty()) {
        auto [cost, x] = pq.top();
        pq.pop();

        if (cost != dist[x])
            continue;

        for (auto &[nxtCost, nxt] : graph[x]) {
            if (dist[nxt] > dist[x] + nxtCost) {
                dist[nxt] = dist[x] + nxtCost;
                pq.push({ dist[nxt], nxt });
            }
        }
    }

    cout << dist[n];
}
